### PR TITLE
Add symbol resolution and auto-completion to rust symbols

### LINKED ISSUE #33: Map gives false error when using a struct as a key instead of a String
Heres the relevent RON set: 
```
(
 tile_object_stacks: (
    tile_object_stacks:{
       (name:"Units"): (
           current_count: 0,
           max_count: 1
           )
      }
 ), 
)
```

I get this error from the plugin at the opening ( for the key value (name:"Units")
> map key or RONTokenType.} expected, got '('

Despite the error the file does load correctly into my program.

### PR DESCRIPTION
**Pull request by flash-freezing-lava**
The pulled repository contains several changes to intellij-ron.
This proposal is meant to discuss upstream pulls of some or all of the changes.

# Autocompletion to rust types and fields
This is the original purpose of the fork. If the rust plugin is installed, the RON plugin searches RON's struct names in rust code and provides references and auto-completion of rust types and fields.

This can only be done under the assumption, that types use the derived implementations. The plugin will not output any errors or warnings, if fields or types don't match the expected ones. This is, to give no false-postives in cases where a customized Deserialize implementation accepts the struct. Neither does it check, if Deserialize is implemented at all for the type. Also, the symbol resolution currently does not consider serde's attributes like `#[serde(rename_all = "UPPERCASE")]` or `#[serde(flatten)]`.

Example:
```rust
#[derive(Serialize, Deserialize)]
struct MyRonStruct {
  foo: u32,
  bar: String,
}
```

```ron
MyRonStruct( // Type is resolved to the rust struct above, and can be followed by Ctrl+B.
  foo: "abc", // Field is resolved. Value not marked an error, even though String was found, when u32 was expected
  text: "def", // Not marked an error, even though the field `text` was not found in MyRonStruct
)
```

# Parser rewrite
In the old parser, if there was an error encountered, it usually resulted in the whole file having no parsed structure at all. The rewrite aims to keep the unparsed part of the file minimal, so the plugin can inspect the error's surroundings in a structured way,
and use its information for symbol resolution and code-completion. The new parser also parses previously unparsable but valid RON structures (like giving `inf` for floats or structs as map keys). This also fixes #33.

The following images compare the current and proposed parsing of a RON file with a complex map key in field `tuple_to_struct`, a missing comma after field `foo`, and auto-completion triggered after the last field.

## Current Version
Current Version

## Proposed Version
Proposed Version

# Tests
I added tests for the newly implemented features (new parser, reference resolution, auto-completion). However, it adds several additional test dependencies, because the tests use [intellij-directory-tests](https://github.com/flash-freezing-lava/intellij-directory-tests), a small [Kotest](https://kotest.io/)-based test framework, I previously wrote for unpublished plugins.

### CODE REVIEW

**Comment by madwareru** [comment]
Wow, this is surely very cool, thanks :)
I will be on a vacation from the next week and I think I could invenst some time to do a review

**Comment by madwareru** [comment]
This PR will lead to a close of #2 potentially

**Comment by madwareru** [comment]
I really like what you did, but there are some problems.
It fails on a test step, please take a look at what happens here:

```
> Task :compileTestKotlin FAILED
FAILURE: Build failed with an exception.
> Task :classpathIndexCleanup
14 actionable tasks: 13 executed, 1 up-to-date

* What went wrong:
Execution failed for task ':compileTestKotlin'.
> Could not resolve all files for configuration ':testCompileClasspath'.
   > Could not find io.github.flash-freezing-lava:intellij-directory-tests:0.1.0.
     Searched in the following locations:
       - https://cache-redirector.jetbrains.com/intellij-dependencies/io/github/flash-freezing-lava/intellij-directory-tests/0.1.0/intellij-directory-tests-0.1.0.pom
       - https://cache-redirector.jetbrains.com/intellij-repository/releases/io/github/flash-freezing-lava/intellij-directory-tests/0.1.0/intellij-directory-tests-0.1.0.pom
       - https://repo.maven.apache.org/maven2/io/github/flash-freezing-lava/intellij-directory-tests/0.1.0/intellij-directory-tests-0.1.0.pom
       - file:/home/runner/.m2/repository/io/github/flash-freezing-lava/intellij-directory-tests/0.1.0/intellij-directory-tests-0.1.0.pom
       - file:/home/runner/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2023.1/79ab2d2970b44e4c36be667e173da87c982a9814/ideaIC-2023.1/io.github.flash-freezing-lava/intellij-directory-tests/0.1.0/ivy-0.1.0.xml
       - file:/home/runner/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2023.1/79ab2d2970b44e4c36be667e173da87c982a9814/ideaIC-2023.1/intellij-directory-tests-0.1.0-2-withKotlin.xml
       - https://cache-redirector.jetbrains.com/plugins.jetbrains.com/maven/io/github/flash-freezing-lava/intellij-directory-tests/0.1.0/intellij-directory-tests-0.1.0.pom
       - file:/home/runner/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/io.github.flash-freezing-lava/intellij-directory-tests-0.1.0-3.xml
     Required by:
         project :
```

**Comment by flash-freezing-lava** [comment]
The problem was, that the test framework isn't published on maven central yet. For experimenting purpose, I fixed it by building it in CI, but will upload artifacts to sonatype's maven repo, if you want to merge the tests.

File: gradle.properties [hunk 1]
```diff
@@ -5,19 +5,22 @@ pluginGroup = com.github.madwareru.intellijronremix
 pluginName = RON Extended Support
 pluginVersion = 0.2.11
 
-pluginSinceBuild = 231.0
-pluginUntilBuild = 231.*
+pluginSinceBuild = 232.0
```
Comment (madwareru) [inline gradle.properties:8]: Please add an entry in CHANGELOG.md about the update of the version of IntelliJ Platform supported

**Review by madwareru — CHANGES_REQUESTED** [review]
There is a minor change needed, after which we could merge the changes

**Comment by madwareru** [comment]
> will upload artifacts to sonatype's maven repo, if you want to merge the tests

Yeah I thing we need tests. This is an old gestalt that I would like to close. So if there is a need to do something for this, feel free to do it

File: gradle.properties (hunk 1 above)
Comment (flash-freezing-lava) [reply gradle.properties:8]: Added

**Comment by flash-freezing-lava** [comment]
I uploaded artifacts to maven central and added the platform note to the CHANGELOG. From my side, we can merge now.

**Review by madwareru — APPROVED** [review]

**Pull request merged**